pkg/model/multiversion/delta.go

```diff
@@ -0,0 +1,257 @@
+	if err != nil {
+		return nil, fmt.Errorf("cannot get resource field renames: %s", err)
+	}
+
+	specDeltas, err := ComputeFieldsDiff(spokeCRD.SpecFields, hubCRD.SpecFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute spec fields deltas: %s", err)
+	}
+
+	statusDeltas, err := ComputeFieldsDiff(spokeCRD.StatusFields, hubCRD.StatusFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute status fields deltas: %s", err)
+	}
+
+	return &CRDDelta{
+		specDeltas,
+		statusDeltas,
+	}, nil
+}
+
+// fieldChangedToSecret returns true if field changed from string to secret.
+func fieldChangedToSecret(src, dst *ackmodel.Field) bool {
+	return (src.FieldConfig == nil ||
+		(src.FieldConfig != nil && !src.FieldConfig.IsSecret)) &&
+		(dst.FieldConfig != nil && dst.FieldConfig.IsSecret)
+}
+
+// ComputeFieldsDiff computes the difference between two maps of fields. It returns a list
+// of FieldDelta's that contains the ChangeType and at least one field reference.
+func ComputeFieldsDiff(
+	spokeCRDFields map[string]*ackmodel.Field,
+	hubCRDFields map[string]*ackmodel.Field,
+	// the hub renames
+	renames map[string]string,
+) ([]FieldDelta, error) {
+	deltas := []FieldDelta{}
+
+	// collect field names and sort them to ensure a determenistic output order.
+	spokeFieldsNames := []string{}
+	for name := range spokeCRDFields {
+		spokeFieldsNames = append(spokeFieldsNames, name)
+	}
+	sort.Strings(spokeFieldsNames)
+
+	hubFieldsNames := []string{}
+	for name := range hubCRDFields {
+		hubFieldsNames = append(hubFieldsNames, name)
+	}
+	sort.Strings(hubFieldsNames)
+
+	// let's make sure we don't visit fields more than once - especially
+	// when fields are renamed.
+	visitedFields := map[string]bool{}
+
+	// first let's loop over the spokeFieldsNames array and see if we can find
+	// the same field name in hubFieldNames.
+	for _, spokeFieldName := range spokeFieldsNames {
+		spokeField, _ := spokeCRDFields[spokeFieldName]
+		hubField, ok := hubCRDFields[spokeFieldName]
+		// If a field is found in both arrays only three changes are possible:
+		// None, TypeChange and ChangeTypeShapeChangedToSecret.
+		// NOTE(a-hilaly): carefull about X -> Y then Z -> X renames. It should
+		// not be allowed.
+		if ok {
+			// mark field as visited.
+			visitedFields[spokeFieldName] = true
+			// check if field change from string to secret
+			if fieldChangedToSecret(spokeField, hubField) {
+				deltas = append(deltas, FieldDelta{
+					Spoke:      spokeField,
+					Hub:        hubField,
+					ChangeType: FieldChangeTypeShapeChangedFromStringToSecret,
+				})
+				continue
+			}
+			// check if field changed from secret to string
+			if fieldChangedToSecret(hubField, spokeField) {
+				deltas = append(deltas, FieldDelta{
+					Spoke:      spokeField,
+					Hub:        hubField,
+					ChangeType: FieldChangeTypeShapeChangedFromSecretToString,
+				})
+				continue
+			}
+
+			// TODO(a-hilaly) better function for equalizing shapes
```

Comment:
s/equalizing/comparing/

Reply:
Removed `TODO` and implemented a proper function to compare two shapes `isEqualShape`